platform/ui/src/contextProviders/ViewportGridProvider.tsx

```diff
@@ -6,120 +6,125 @@ import React, {
+  layout: Layout;
+  viewports: Map<string, Viewport>;
+}
+
+const DEFAULT_STATE: DefaultState = {
+  activeViewportId: null,
   layout: {
     numRows: 0,
     numCols: 0,
     layoutType: 'grid',
   },
-  viewports: [
-    {
-      displaySetInstanceUIDs: [],
-      viewportOptions: {},
-      displaySetSelectors: [],
-      displaySetOptions: [{}],
-      x: 0, // left
-      y: 0, // top
-      width: 100,
-      height: 100,
-      viewportLabel: null,
-    },
-  ],
+  // Viewports structure has been changed to Map (previously it was
+  // tied to the viewportIndex which caused multiple issues. Now we have
+  // moved completely to viewportId which is unique for each viewport.
+  viewports: new Map(
+    Object.entries({
+      default: {
+        viewportId: 'default',
+        displaySetInstanceUIDs: [],
+        viewportOptions: {},
+        displaySetSelectors: [],
+        displaySetOptions: [{}],
+        x: 0, // left
+        y: 0, // top
+        width: 100,
+        height: 100,
+        viewportLabel: null,
+      },
+    })
+  ),
 };
 
-export const ViewportGridContext = createContext(DEFAULT_STATE);
-
-/** A viewport is reuseable if it is the same size as the old
- * one and has the same display sets, in the same position.
- * It SHOULD be possible to re-use them at different positions, but
- * this causes problems with segmentation.
- */
-const isReuseableViewport = (oldViewport, newViewport) => {
-  const sameDiplaySets = isEqual(
-    oldViewport.displaySetInstanceUIDs,
-    newViewport.displaySetInstanceUIDs
-  );
-  return (
-    oldViewport.viewportIndex === newViewport.viewportIndex &&
-    sameDiplaySets &&
-    oldViewport.height === newViewport.height &&
-    oldViewport.width === newViewport.width
-  );
+const getViewportLabel = (viewports, viewportId) => {
+  const viewportIds = Array.from(viewports.keys());
+  return viewportLabels[viewportIds.indexOf(viewportId)];
 };
 
-// Holds a global viewport counter - used to assign new id's to viewports
-// Starts at a value above zero so that any of the old viewport id's are
-// immediately obvious and if we get any index generated viewports, they are
-// definitely distinct from these ones.
-let viewportCounter = 5000;
-
-/**
- * Find a viewportId to re-use if possible, preserving the existing
- * viewport information, OR create a new one if the viewport given isn't
- * compatible with what was there before.
- *
- * @param viewportIdSet
- * @param viewport
- * @param stateViewports
- * @returns
- */
-const reuseViewportId = (viewportIdSet: Set, viewport, stateViewports) => {
-  for (const oldViewport of stateViewports) {
-    const { viewportId: oldId } = oldViewport;
-    if (!oldId) {
-      // This occurs on startup, so skip re-using it
-      continue;
-    }
-    if (viewportIdSet.has(oldId)) {
-      // oldId is already used - we can't reuse it
-      continue;
+const getActiveViewportId = (state: DefaultState, newViewports: Map) => {
```

Comment:
This kinda confuses me with the `state.activeViewportId`. Please add JS doc comments for it and consider renaming the function to something like `determineActiveViewportId` (I don't have a good name, but the current name confuses me a bit).